mxnet_model_service.py

```diff
@@ -0,0 +1,95 @@
+import mxnet as mx
+import numpy as np
+import tarfile
+import json
+import os
+
+from mxnet.gluon.utils import download
+from mxnet.io import DataBatch
+from model_service import SingleNodeModel, MultiNodesModel
+
+URL_PREFIX = ('http://', 'https://', 's3://')
+MODEL_DIR = 'models/'
+CONFIG_FILE = 'config.json'
+
+
+def check_input_shape(inputs, metadata):
+    """Check input data shape consistency with metadata.
+
+    Parameters
+    ----------
+    inputs : List of NDArray
+        Input data in NDArray format.
+    metadata : dict
+        Dictionary containing model metadata.
+    """
+    data_names = metadata['input'].keys()
+    data_shapes = metadata['input'].values()
+    assert isinstance(inputs, list), "Input data must be a list."
+    assert len(inputs) == len(data_names), "Input number mismatches with " \
+                                           "metadata. %d expected but got %d." \
+                                           % (len(data_names), len(inputs))
+    for input, data_shape, data_name in zip(inputs, data_shapes, data_names):
+        assert isinstance(input, mx.nd.NDArray), "Each input must be NDArray."
+        assert len(input.shape) == \
+               len(data_shape), "Shape dimension of input %s mismatches with " \
+                                "metadata. %d expected but got %d." \
+                                % (data_name, data_shape,
+                                   len(input.shape))
+        for idx in range(len(input.shape)):
+            if idx != 0 and data_shape[idx] != 0:
```

Comment:
What does this mean? when skip idx = 0? when can be data_shape[idx] is zero?

Reply:
Since the first axis of shape is batch size, so we skip it.
In MXNet, 0 in shape means this axis can be any number.